Private chat mode

AKMDM opened the pull request:
This adds a new chat mode to Zandronum, allowing the player (or the server) to send chat messages privately to another player. Nobody else can see these messages and they won't be logged into the server's chat log, unless the server is sending or receiving a private message.

TorrSamaho commented on src/c_console.cpp:
```diff
@@ -285,6 +286,12 @@ CUSTOM_CVAR (Int, msg4color, 0, CVAR_ARCHIVE)
 
+CUSTOM_CVAR (Int, msg5color, 21, CVAR_ARCHIVE)
+{
+	// [AK] Private chat message color.
```
Please move this comment to where the delta to GZDoom starts, i.e. before `CUSTOM_CVAR (Int, msg5color ....`

TorrSamaho commented on src/cl_commands.cpp:
```diff
@@ -282,6 +282,22 @@ void CLIENTCOMMANDS_Say( ULONG ulMode, const char *pszString )
 
+//*****************************************************************************
+//
+void CLIENTCOMMANDS_PrivateSay( ULONG ulPlayer, const char *pszString )
```
This function shares a lot of code with `CLIENTCOMMANDS_Say`. I think it would be better if `CLIENTCOMMANDS_Say` is extended to cover the case that `ulMode` is `CHATMODE_PRIVATE_SEND`. For this it would need an additional optional argument `ulPlayer` that defaults to `MAXPLAYERS`.

TorrSamaho commented on src/doomtype.h:
```diff
@@ -151,7 +153,8 @@ enum
 #define PRINT_TEAMCHAT			4				// chat messages from a teammate
-#define PRINT_LOG				5				// only to logfile
+#define PRINT_PRIVATECHAT		5				// private chat messages from another player
+#define PRINT_LOG				6				// only to logfile
```
This delta to GZDoom code should be marked with a comment. I'd also not change `PRINT_LOG`, since that is from GZDoom. Or is it necessary that `PRINT_PRIVATECHAT` is `PRINT_TEAMCHAT+1`?

AKMDM replied:
I figured private chat messages would have greater importance than team chat messages, so it should be one level higher. Would changing `PRINT_PRIVATECHAT` so that it matched `PRINT_LOG 5` be acceptable?

TorrSamaho replied:
Ah, I see. In that case it's fine to raise `PRINT_LOG` to `6`. Just leave comments to mark that you added `PRINT_PRIVATECHAT` and that you raised `PRINT_LOG` to make room.

TorrSamaho commented on src/sv_commands.cpp:
```diff
@@ -1061,6 +1061,33 @@ void SERVERCOMMANDS_PlayerSay( ULONG ulPlayer, const char *pszString, ULONG ulMo
+	}
+
+	// [AK] Next, send the command back to the player who sent the message.
+	if ( ulSender != MAXPLAYERS )
```
This will happen even if the message was not send due to `(( ulReceiver != MAXPLAYERS ) && ( !bForbidChatToPlayers || players[ulReceiver].bSpectating ))` being false. I think in that case, the sender should receive an error.

AKMDM commented:
I'm finishing up on these adjustments, if that's everything. I should have a new commit ready very soon.

EDIT: Just committed the adjustments, I hope these are okay. One more thing, I added extra code to allow the client to cycle backwards through the player list if they're pressing TAB + SHIFT. Otherwise, it cycles forward with just TAB.

TorrSamaho commented:
> I'm finishing up on these adjustments, if that's everything. I should have a new commit ready very soon.


Yes, I'm done checking and commented on everything I noticed.

TorrSamaho commented on src/chat.cpp:
```diff
@@ -609,17 +609,23 @@ bool CHAT_Input( event_t *pEvent )
 				{
-					ULONG oldPlayer = g_ulChatPlayer;
+					int tempPlayer = g_ulChatPlayer;
+					int direction = ( pEvent->data3 & GKM_SHIFT ) ? -1 : 1;
```
Are these changes here related to any of the points I brought up?  At first glance, it doesn't look like this to me. I also just realize that github seems to have hid the changes to chat.cpp when I made the review.

AKMDM replied:
This particular change isn't related to what you mentioned last week, it's just something I forgot to add when I originally made the pull request.

Originally, the client could only move forwards through the player list if they wanted to change who to send the private message to, whenever they pressed TAB. Now, they can move backwards if they press TAB + SHIFT at the same time. This is useful in case the client were to accidentally pass the player they intend to privately chat with and need to go backward/forward.

TorrSamaho replied:
I see, sounds like a nice addition.

TorrSamaho commented on src/chat.cpp:
```diff
@@ -594,7 +607,30 @@ bool CHAT_Input( event_t *pEvent )
 			else if ( pEvent->data1 == '\t' )
 			{
-				g_ChatBuffer.TabComplete();
+				if (( g_ulChatMode == CHATMODE_PRIVATE_SEND ) && ( g_ulChatPlayer != MAXPLAYERS ))
```
I think it would be safer to also check `g_ulChatPlayer != consoleplayer` and `SERVER_CalcNumPlayers() >= 2`. In case both are false, the while loop won't terminate.

AKMDM replied:
I agree, I will add this check as well.

TorrSamaho commented on src/chat.cpp:
```diff
@@ -594,7 +607,30 @@ bool CHAT_Input( event_t *pEvent )
+						if ( tempPlayer == g_ulChatPlayer )
+							break;
+					}
+					while (( PLAYER_IsValidPlayer( tempPlayer ) == false ) || ( tempPlayer == consoleplayer ));
```
Perhaps you want to skip bots? In that case, you need to check `SERVER_CountPlayers( false ) >= 2` instead of `SERVER_CalcNumPlayers() >= 2` in the comment above.

EDIT: I just noticed that we cannot select the server with this. Wouldn't it be more user friendly if we'd also allow the server to by cycled through? This way one also doesn't need to count the players to prevent an infinite loop.

AKMDM replied:
Maybe so. Perhaps the client should also only be allowed to use the private chat if the network state is `NETSTATE_CLIENT`? There might not be any real reason for the player to send private messages in a singleplayer game.

TorrSamaho replied:
Good point. I agree that there is no need for private chat in single player or when watching demos.

TorrSamaho commented on src/chat.cpp:
```diff
@@ -1131,6 +1225,152 @@ CCMD( say_team )
+
+	if ( argv.argc() >= 2 )
+	{
+		if ( stricmp( argv[1], "Server" ) == 0 )
```
So if a player has the name "Server", this player will not be reachable by `sayto`, right? I don't think that this is really a problem, but one should just be aware of this.

AKMDM replied:
Technically yes, since it first assumes the client is sending a private message to the server rather than a player who may have the same name. I think it's unlikely somebody will try to use a name exactly like that (I remember some hosts would even forbid players from using it to prevent impersonation), but it's still possible. Should a player name exactly like "Server" even be allowed?

TorrSamaho commented on src/chat.cpp:
```diff
@@ -1131,6 +1225,152 @@ CCMD( say_team )
+		if ( argv.SafeGetNumber( 1, playerIndex ) == false )
+			return;
+
+		if ( PLAYER_IsValidPlayer( playerIndex ) == false )
```
Should we allow the server to be contacted with idx "-1"?

AKMDM replied:
Yes, this should be allowed too.

TorrSamaho commented on src/chat.cpp:
```diff
@@ -1131,6 +1225,152 @@ CCMD( say_team )
+				return false;
+			}
+		}
+		while (( PLAYER_IsValidPlayer( g_ulChatPlayer ) == false ) || ( g_ulChatPlayer == (ULONG)consoleplayer ));
```
Looks like this will end in an infinite loop of there is only one player on the server.

TorrSamaho commented on src/chat.cpp:
```diff
@@ -1131,6 +1225,152 @@ CCMD( say_team )
+//
+// [AK] Allows players (or the server) to send private messages to other players.
+//
+bool chat_CanSendPrivateMessage( ULONG ulPlayer )
```
The name of the function sounds as if it would only check whether `ulPlayer` is allowed to send a private message, but what it actually does it to change the global variable `g_ulChatPlayer` to a valid private chat target. The name of the function should reflect that. I would also rename the argument `ulPlayer` to `ulSender`.

TorrSamaho commented on src/chat.cpp:
```diff
@@ -1131,6 +1225,152 @@ CCMD( say_team )
+//
+bool chat_CanSendPrivateMessage( ULONG ulPlayer )
+{
+	if ( ulPlayer == MAXPLAYERS )
```
We don't need to make sure g_ulChatPlayer is a valid target if the sender is the server?

AKMDM replied:
You're right, that needs to checked as well.

TorrSamaho commented:
Sorry for overlooking the changes in `chat.cpp` last time. I didn't realize that github automatically hides changed files with too many changes...

AKMDM commented:
No worries! I'm glad we took a good look in this file to see where things need to be improved for next time. I'll probably have these changes ready for next week, if that's okay. Thanks for the feedback!

TorrSamaho commented:
Sure, there's no rush. Have a nice week!

AKMDM commented:
Thank you, you as well. :)

AKMDM commented:
I submitted a couple of new commits.

- I fixed the private chat mode so that it now only works in online games and bots cannot be selected. Also, if the client tries to start typing a private message when there's no other clients, the server is selected instead.
- `chat_CanSendPrivateMessage( ULONG ulPlayer )` is replaced with `chat_IsPlayerValidReceiver( ULONG ulPlayer )` and `chat_FindValidReciever ( void )`. The former checks if a player is eligible to become a receiver of a private message, and the latter checks for a valid player that can be a receiver. Otherwise, the server is chosen.
- The possibility of any infinite loops occurring in the code should be resolved now.
- The server can now be selected by passing `-1` to `sayto_idx`.

I also added more commits, some of which aren't necessarily related to the private chat:

- Players cannot name themselves as "server" in online games anymore. This should avert potential impersonation of the server, and the possibility that someone tries using `sayto "server"` when there's a player with the same name.
- The chat cursor now blinks while the playing is creating a message. If they're typing or passing any input, the cursor remains visible. The additions are somewhat complicated, but I had to make sure `displayString` still looks convincing on the screen while the cursor is supposed to be invisible by taking account the width of the '_' character.
- I touched up and re-added my commit which adds the new EVENT type: `GAMEEVENT_CHAT` and the ACS function `GetChatMessage( int player, int index )`. The biggest difference is that up to three chat messages from the player (or server) are stored in memory, so modders can also pass an `index` between 0 and 2 to get that corresponding chat message, with ~~0 being the oldest and 2 being newest~~. This is in case we receive more than one chat message from the client in a single tic, in which the last chat message replaces the older one.

EDIT: An index of 0 returns the newest chat message, and 2 returns the oldest chat message.

TorrSamaho commented on src/chat.cpp:
```diff
@@ -508,6 +520,18 @@ void CHAT_Tick( void )
+	// they're in the game. If not, cancel the message entirely.
+	if ( g_ulChatMode == CHATMODE_PRIVATE_SEND )
+	{
+		if (( g_ulChatPlayer != MAXPLAYERS ) && ( PLAYER_IsValidPlayer( g_ulChatPlayer ) == false ))
```
Shouldn't this should be using `chat_IsPlayerValidReceiver` instead of `PLAYER_IsValidPlayer` now?

AKMDM replied:
I must've overlooked this line, sorry about that.

TorrSamaho commented on src/chat.cpp:
```diff
@@ -594,7 +618,30 @@ bool CHAT_Input( event_t *pEvent )
 			else if ( pEvent->data1 == '\t' )
 			{
-				g_ChatBuffer.TabComplete();
+				if (( g_ulChatMode == CHATMODE_PRIVATE_SEND ) && ( SERVER_CountPlayers( false ) >= 2 ))
```
This is also breaking now when `tempPlayer == MAXPLAYERS `, so there is no need anymore to check `SERVER_CountPlayers( false ) >= 2`, right?

AKMDM replied:
`SERVER_CountPlayers( false ) >= 2` is still needed, since there's no reason to execute the loop and try finding another client who we can privately chat with if none exist. However, the server (MAXPLAYERS) is always a valid receiver, so in this case, we don't really need to check if `tempPlayer` is a valid receiver using `chat_IsPlayerValidReceiver( tempPlayer )`. I hope this makes sense.

EDIT: Looking at it, I probably should've added an `if ( ulPlayer == MAXPLAYERS )` check to `chat_IsPlayerValidReceiver`, it could've simplified the code in a few places.

TorrSamaho replied:
Good point.

TorrSamaho commented:
Thanks for the new patches! However, to keep things more manageable, can you do me a favor and remove the separate commits cd80940, d2fc101, f976178 from this merge request? Let's first get the private chat system merged, before extending it or adding other chat features.

AKMDM commented:
Sure, will do.

~~EDIT: Is it okay if I open up a new pull request? I'm having trouble removing those few commits on my end and ended up making some mistakes~~.

AKMDM commented:
I managed to fix it, Torr.

TorrSamaho commented on src/chat.cpp:
```diff
@@ -1131,6 +1226,215 @@ CCMD( say_team )
+// [AK] Allows players (or the server) to send private messages to other players.
+//
+void chat_PrivateMessage( FCommandLine &argv, const ULONG ulReceiver, const bool bServerSelected )
+{
```
Why do we need a separate argument `bServerSelected` to encode the server as target instead of just using `ulReceiver == MAXPLAYERS`?
BTW: Sorry that this comment only shows up now. I posted it last week but unfortunately using "Start a review" instead of "Add single comment". I just now realized that you couldn't see it.

AKMDM replied:
Thank you for asking. I added `bServerSelected` to ignore all the extra checks required to validate another player being the receiver, if the client explicitly chooses to the server. If the client wants indicates that they want to send a private message to the server using `sayto server` or `sayto_idx -1`, then it's not necessary to check if the receiver is invalid, a bot, ourselves, etc.

Additionally, if `bServerSelected` is true, then the function `chat_FindValidReceiver` won't be entered at all. If we did enter that function and there was only one client connected, then the message "There's no other clients to send private messages to. The server will be selected instead." would always appear in the console, even if the client intentionally chose to send a private message to the server.

TorrSamaho replied:
You could also bypass all those checks by using `ulReceiver == MAXPLAYERS` instead of `bServerSelected`. For instance, you could add `const bool bServerSelected = (ulReceiver == MAXPLAYERS);` in the first line of the implementation of `chat_PrivateMessage` and keep the rest of `chat_PrivateMessage` untouched (except for removing the argument `bServerSelected`). Perhaps there is a reason why you don't want to encode the information `bServerSelected` in the argument `ulReceiver`, but right now I don't see it.

AKMDM replied:
Fair enough, I can change that quickly and resubmit it if you're still around.

TorrSamaho replied:
Yes, I'm still around.

TorrSamaho replied:
Sorry, but I gotta go now. I should be able to add this tomorrow though.

AKMDM replied:
No worries, sorry for the wait. I just around to adding that last patch, but if you can still add it in tomorrow that's awesome!

TorrSamaho commented:
> I managed to fix it, Torr.

Thanks! Looking all good now, except for the one comment I didn't properly post last week. Sorry for that.

TorrSamaho commented on src/chat.cpp:
```diff
@@ -1294,10 +1294,11 @@ void chat_FindValidReceiver( void )
 {
 	ULONG		ulIdx;
 	FString		ChatString;
+	const bool	bServerSelected = ulReceiver;
```
Shouldn't this be `const bool bServerSelected = (ulReceiver == MAXPLAYERS);`?

EDIT: Now I'm really gone for today ;).

AKMDM replied:
Yes, I forgot to fix that (really stupid mistake). :(

EDIT: Uploaded the now fixed commit, so sorry about that!

TorrSamaho replied:
Thanks!

TorrSamaho commented:
I compressed, rebased and added this patch.
